### PR TITLE
Support file specific linter config

### PR DESCRIPTION
**Pull request by Atry**
The file specific linter config is used in #417

### CODE REVIEW

File: tests/LinterConfigTest.hack [hunk 1]
```diff
@@ -87,6 +93,29 @@ final class LinterConfigTest extends HackTest {
     )->toThrow(\Exception::class, 'is not of the correct type');
   }
 
+  public async function testInvalidConfigForCLI(): Awaitable<void> {
+    list($cli, $_, $stdout, $stderr) =
+      $this->getCLIInDirectory(vec[], self::TEST_DIRECTORY);
+    $exit_code = await $cli->mainAsync();
+    expect($stderr->getBuffer())->toMatchRegExp(
+      re'/A linter threw an exception\\:/',
+    );
```
Comment (lexidor) [inline tests/LinterConfigTest.hack:102]: I see you use toMatchRegExp everywhere. If you use toContainSubString you can save yourself (and future readers) from having the reason about escaped regular expression characters.

**Comment by fredemmott** [comment]
This is  probably a worthwhile change on itself, but I don't think it should be needed here: does `/*HHAST_IGNORE_ALL[1234]*/` not work for these? If not, that should be fixed:
- for consistency
- per-file suppressions via config is not scalable, especially as json's lack of support for trailing commas increases the risk of merge conflicts

**Comment by lexidor** [comment]
> This is probably a worthwhile change on itself, but I don't think it should be needed here: does `/*HHAST_IGNORE_ALL[1234]*/` not work for these? If not, that should be fixed:

I believe this PR intends to address my comment on #409

This might also be a nice alternative to `.LICENSE_HEADER.hh.txt`. The LicenseHeaderLinter predates linter configuration and uses a well-known naming convention for setting per directory license headers instead. Once configuration is able to be declared in a override, this system can be supplemented using a configuration.

**Comment by Atry** [comment]
> This is probably a worthwhile change on itself, but I don't think it should be needed here: does `/*HHAST_IGNORE_ALL[1234]*/` not work for these? If not, that should be fixed:
> 
> * for consistency
> * per-file suppressions via config is not scalable, especially as json's lack of support for trailing commas increases the risk of merge conflicts

I suppose people could use this PR to suppress linter issues in dependencies, where you cannot easily add `HHAST_IGNORE_ALL` comments.

**Comment by fredemmott** [comment]
> I suppose people could use this PR to suppress linter errors in dependencies, where you cannot easily add HHAST_IGNORE_ALL comments.

People should not be linting their dependencies; lint-cleanness should be considered an implementation detail

**Comment by fredemmott** [comment]
> This is probably a worthwhile change on itself,

Just to clarify what I meant here:

I think this PR is good, and should probably be merged

I don't think the rest of the stack should depend on it :)

**Comment by Atry** [comment]
Rebased this PR to avoid dependencies between other PRs

File: hhast-lint.json [hunk 2]
```diff
@@ -24,6 +24,14 @@
       "extraLinters": [
         "Facebook\\HHAST\\DataProviderTypesLinter"
       ]
+    },
+    {
+      "patterns": ["src/Migrations/*"],
+      "linterConfigs": {
+        "Facebook\\HHAST\\HHClientLinter": {
+          "ignore": [5624, 5639]
```
Comment (fredemmott) [inline hhast-lint.json:32]: Filed #416 for removing this

File: src/Migrations/AddXHPChildrenDeclarationMethodMigration.hack [hunk 3]
```diff
@@ -43,19 +43,14 @@ final class AddXHPChildrenDeclarationMethodMigration
...
-    if ($t->getLeading() !== $t->getLeadingWhitespace()) {
-      $ut = $use->getFirstTokenx();
-      return $in->withDeclarations(
-        $in->getDeclarations()->insertBefore(
-          $before,
-          $use->replace($ut, $ut->withLeading($t->getLeading())),
-        ),
-      )
-        ->replace($t, $t->withLeading(null));
-    }
```
Comment (fredemmott) [inline src/Migrations/AddXHPChildrenDeclarationMethodMigration.hack:55]: Doesn't look like these changes are meant to be in this PR?

File: src/__Private/LintRunConfig.hack [hunk 4]
```diff
@@ -39,21 +51,10 @@ final class LintRunConfig {
...
-        ?'disabledAutoFixes' => vec<string>,
-        ?'disableAllAutoFixes' => bool,
-        ?'disableAllLinters' => bool,
-      )
-    >,
+    ?'overrides' => vec<self::TOverride>,
     // Each linter may specify a type for itself.
     // The type of this key is effectively `dict<classname<T1 ... Tn>, Tx>`
-    ?'linterConfigs' => dict<string, dynamic>,
+    ?'linterConfigs' => dict<string, shape(...)>,
```
Comment (fredemmott) [inline src/__Private/LintRunConfig.hack:57]: revert to dynamic, and use dynamic for the new option

https://github.com/hhvm/hhast/pull/364#discussion_r707531021

**Review by fredemmott — APPROVED** [review]
use dynamic instead of shape and remove XHP changes before merging

**Comment by fredemmott** [comment]
XHP change seems needed because of this, but should be a separate PR (and separate commit in the main branch) as it's independent and should be reviewed (and potentially bisected or reverted) separately

File: src/Migrations/AddXHPChildrenDeclarationMethodMigration.hack (hunk 3 above)
Comment (Atry) [reply src/Migrations/AddXHPChildrenDeclarationMethodMigration.hack:55]: This change is to fix a 5607 linter error. I reviewed `$t->getLeading() !== $t->getLeadingWhitespace()` and I believe the condition is always true, therefore, I think the `if` clause is unnecessary.

Comment (Atry) [reply src/Migrations/AddXHPChildrenDeclarationMethodMigration.hack:55]: Anyway, I will split the change to another PR

**Pull request merged**